pkg/uio/reader_test.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uio
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func readAndCheck(t *testing.T, want, tmpfileP string) {
+	t.Helper()
+	r := strings.NewReader(want)
+	if err := ReadIntoFile(r, tmpfileP); err != nil {
+		t.Errorf("ReadIntoFile(%v, %s) = %v, want no error", r, tmpfileP, err)
+	}
+
+	got, err := ioutil.ReadFile(tmpfileP)
+	if err != nil {
+		t.Fatalf("ioutil.ReadFile(%s) = %v, want no error", tmpfileP, err)
+	}
+	if want != string(got) {
+		t.Errorf("got: %v, want %s", string(got), want)
+	}
+}
+
+func TestReadIntoFile(t *testing.T) {
+	want := "I am the wanted"
+
+	dir, err := os.MkdirTemp("", "uio-reader-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	// Write to a file already exist.
+	tmpfile, err := ioutil.TempFile(dir, "uio-out")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := tmpfile.Close(); err != nil {
+		t.Fatal(err)
+	}
+	readAndCheck(t, want, tmpfile.Name())
+
+	// Write to a file that does not exist.
+	p := filepath.Join(dir, "uio-out-not-existing")
+	readAndCheck(t, want, p)
+
+	// Write to an existing file that has pre-existing content.
+	tmpfile, err = ioutil.TempFile(dir, "uio-out-prexist-content")
```

Comment:
where did you create this file before? If you use os.WriteFile, you could just do it again, replace lines 55-64 with
ioutil.WriteFile(filepath.Join(dir, "uio-out"), []byte(whatever))

Also, why not on line 59 use want, not that []byte("...")?

Reply:
Done. 

this is a newly created file, right here :). 

I could use want as pre-existing content, but I then don't know what I checked is newly written or from pre-existing content

Reply:
You have a temporary directory, created via TempDir, and in the directory you can make files. If you have TempDir, which will have a unique name, since it is partly randomized, you do not need to call TempFile, since you can create files in the directory you got from TempDir. 

TempFile is useful when you are using a temporary file in a shared directory such as /tmp, as it makes the effort to create  a name that will not collide with other names. That's not a problem here. You created a temporary directory, with TempDir, and its path is dir. To create other files, you do not need to use TempFile. You can, for example, do things like this
pre := filepath.Join(dir, "uio-out-prexist-content")
and then use the name pre with ioutil.WriteFile, os.OpenFile, and so on. 

Again, there is no need to use TempFile. The main function of TempFile is to create a file with a partially random name to avoid collisions, which can happen in a common directory such as /tmp. By using TempDir, which creates a uniquely named directory, you avoid the need to call TempFile. Just create your files in the directory you created with TempDir.

Further, since you are trying to do tests with file names that do not exist, it is best to use TempDir to ensure that you can create file names that really do not exist. That is why using TempDir is a good approach.

Reply:
Done - removed tempFilr, and using OpenFile instead.

I saw the point on using `TempDir`, and I get the point of using `TempFile` within a `TempDir` can be unnecessary. I was just thinking that having a randomized tempfile under a TempDir does not hurt :) -- But sounds like that is not conventional, and may reads quirky to readers who see the code later!

Reply:
Looks like golang used in Circle CI is not greater or equal than `1.16`, and `os.MkdirTemp` is not supported ...

```
#!/bin/bash -eo pipefail
go vet -composites=false ./cmds/... ./pkg/... ./

# github.com/u-root/u-root/pkg/uio
vet: pkg/uio/reader_test.go:34:17: MkdirTemp not declared by package os

Exited with code exit status 2
CircleCI received exit code 2
```

Reply:
I guess we will need to look into bumping the version, or resort back to TempFile at all

Reply:
resolved offline.